src/libsyntax/feature_gate.rs

```diff
@@ -718,6 +719,41 @@ pub enum AttributeGate {
     Ungated,
 }
 
+#[derive(Clone, Copy)]
+pub struct AttributeTemplate {
+    word: bool,
+    list: Option<&'static str>,
+    name_value_str: Option<&'static str>,
+}
+
+impl AttributeTemplate {
+    fn compatible(&self, meta_item_kind: &ast::MetaItemKind) -> bool {
+        match meta_item_kind {
+            ast::MetaItemKind::Word => self.word,
+            ast::MetaItemKind::List(..) => self.list.is_some(),
+            ast::MetaItemKind::NameValue(lit) if lit.node.is_str() => self.name_value_str.is_some(),
+            ast::MetaItemKind::NameValue(..) => false,
+        }
+    }
+}
+
+macro_rules! template {
```

Comment:
This is nice, but it'd be even nicer to have some docs kind of explaining what it's purpose is and maybe giving a few examples? Or maybe scatter the docs in the macro arms below -- something to help guide future people editing this file and trying to add something =)